Elaborate on filesystem socket permission issues

rlbdv opened the pull request:
I wondered if we might want to elaborate a bit on the filesystem socket permission constraints, also discussed in Linux's `unix(7)`.  I also didn't know if having non-ascii "dashes" in the `start-server` docstring was intentional, but I hadn't noticed before, and ao changed `:socket` to match the others.

From the manpage:
```
Pathname socket ownership and permissions
    In  the  Linux  implementation,  pathname sockets honor the permissions of the directory
    they are in.  Creation of a new socket fails if the process  does  not  have  write  and
    search (execute) permission on the directory in which the socket is created.

    On Linux, connecting to a stream socket object requires write permission on that socket;
    sending a datagram to a datagram socket  likewise  requires  write  permission  on  that
    socket.   POSIX  does  not  make  any statement about the effect of the permissions on a
    socket file, and on some systems (e.g., older BSDs), the socket permissions are ignored.
    Portable programs should not rely on this feature for security.

rlbdv commented:
Test failures appear to be the reflection warnings again.  I thought we'd resolved all those, but while `lein check` is fine, it looks like eastwood is not.  I don't know if that's new, or if I was only working from `lein check` before, but I did notice a new version of easwood being fetched when I ran the tests locally.  Please let me know how you'd prefer I try to handle that.  Thanks.

bbatsov commented:
I'm fine with whatever solution for the reflection warning you prefer. I think @vemv mentioned something on the subject in a different PR. (#239)

rlbdv commented:
Well, the `(set! *warn-on-reflection* false)` "works" for `lein check` but doesn't work for (current?) eastwood, and I don't know what will, but I can try (again) to figure something else out.  I believe the difficulty is that we can't mention classes that might or might not be available in type annotations as far as I can tell right now, and eastwood doesn't respect that `set!` override.

And really, we *can't* legitimately get rid of reflection warnings in these cases?  That is, the classes really are unknowable at compile time.

I can imagine some other hacks, but I don't know if they'll work, or if you'd want them.  i.e. if eastwood will let us opt-out of the checks for a particular ns or function, the we could segregate the few bits that are affected.  (Note that I suspect the complaints may vary depending on the JDK version and/or use of `+junixsocket` since that changes the available classes.)

Also, fwiw, this particular PR is just an adjustment to the docs, though I'm happy to help fix the broader issue if we can figure out how we want to.

vemv commented:
The relevant discussion is https://github.com/nrepl/nrepl/pull/204#discussion_r675346773 , ideally @bbatsov would chime in

In any case reflection warnings, like any other warning, can be silenced.

See https://github.com/nrepl/nrepl/blob/f7d489f4027671b78713a033b97aae84b57d5858/project.clj#L86-L87 

and https://github.com/jonase/eastwood#ignored-faults

rlbdv commented:
Ahh, ok, thanks.  I'll take a look.

rlbdv commented:
@bbatsov I've pushed an update with one potential solution (see the two additional commits).  I'm also happy to break that out into a separate pr if that'd be preferable.  We could also use eastwood's "line number" based suppression instead, but I wasn't sure you'd want to have to update the suppression every time the line number changes (FWIW I've also suggested adding function name suppressions to eastwood https://github.com/jonase/eastwood/issues/415 which would avoid the need for the separate namespace.)

And if y'all would prefer the approach using dynamic annotations via macroexpansion, that's fine too.  Personally, I still lean toward avoiding the complexity/risk there on balance, but happy to change this to work that way if you like.

vemv reviewed (approved):
LGTM

I'd humbly suggest, please remember about https://github.com/nrepl/nrepl/pull/204#discussion_r682477225 - it's not good to have code that isn't exercised in CI in any way.

Sibling projects like Orchard have had problems in the past due to insufficient test coverage. For dense JDK stuff it's better to have the author in question provide enough assurance in advance

rlbdv commented:
I removed the CI clj-kondo install (which should be obsolete now -- see commit message), and added junixsocket testing (cf. https://github.com/nrepl/nrepl/pull/204#discussion_r684800064).  For discussion, I just added it to `make test`, which means that it'll run for every clj/jdk combination, which will approximately double the CI run times, and which might not be deemed necessary.

Not sure, but happy to adjust that as you like.  Also happy to break some of all of this out into separate prs if that'd be preferred.

bbatsov commented:
Looks good to me as well. Thanks!
